trie/stacktrie_test.go

```diff
@@ -240,3 +240,22 @@ func TestDerivableList(t *testing.T) {
 		}
 	}
 }
+
+func TestUpdate(t *testing.T) {
+	st := NewStackTrie(nil)
+	nt, _ := New(common.Hash{}, NewDatabase(memorydb.New()))
+	kvs := []struct {
+		K string
+		V string
+	}{
+		{"63303030", "3041"}, // stacktrie.Update
+		{"65", "3000"},       // stacktrie.Update
+	}
+	for _, kv := range kvs {
+		nt.TryUpdate(common.FromHex(kv.K), common.FromHex(kv.V))
+		st.TryUpdate(common.FromHex(kv.K), common.FromHex(kv.V))
+	}
+	if nt.Hash() != st.Hash() {
+		t.Fatalf("error %x != %x", st.Hash(), nt.Hash())
+	}
+}
```

Comment:
I figured out the issue in the stacktrie!

It's wrong here https://github.com/ethereum/go-ethereum/blob/master/trie/stacktrie.go#L327
The RLP input should be the struct like

```go
var valuenode node
if len(st.children[0].val) < 32 {
	valuenode = rawNode(st.children[0].val)
} else {
	valuenode = hashNode(st.children[0].val)
}
n := struct {
	Key []byte
	Val node
}{
	Key: hexToCompact(st.key),
	Val: valuenode,
}
```

Reply:
If the child of the extension node is the embedded node(size less than 32), the type of value is "RLPed rawvalue". We shouldn't RLP the slice again.